packages/devtools_app/lib/src/eval_on_dart_library.dart

```diff
@@ -156,25 +178,229 @@ class EvalOnDartLibrary {
+      scope: {'widgetInspectorService': widgetInspectorServiceRef.id},
+    ).then((ref) => ref.valueAsString);
+
+    await safeEval(
+      '() async {'
+      '  final reader = widgetInspectorService.toObject("$readerId", "$readerGroup") as List;'
+      '  try {'
+      '    final result = $expression;'
+      '    reader.add(result);'
+      '  } catch (err, stack) {'
+      '    reader.add(err);'
+      '    reader.add(stack);'
+      '  } finally {'
+      '    postEvent("future_completed", {"id": $id});'
+      '  }'
+      '}()',
+      isAlive: isAlive,
+      scope: {
+        ...?scope,
+        'postEvent': postEventRef.id,
+        'widgetInspectorService': widgetInspectorServiceRef.id,
+      },
+    );
+
+    await future;
+
+    final resultRef = await evalInstance(
+      '() {'
+      '  final result = widgetInspectorService.toObject("$readerId", "$readerGroup") as List;'
+      '  widgetInspectorService.disposeGroup("$readerGroup");'
+      '  return result;'
+      '}()',
+      isAlive: isAlive,
+      scope: {'widgetInspectorService': widgetInspectorServiceRef.id},
+    );
+
+    assert(resultRef.length == 1 || resultRef.length == 2);
+    if (resultRef.length == 2) {
+      throw FutureFailedException(
+        expression,
+        resultRef.elements[0],
+        resultRef.elements[1],
+      );
+    }
+
+    return resultRef.elements[0];
+  }
+
+  /// An [eval] that throws when a [Sentinel]/error occurs or if [isAlive] was
+  /// disposed while the request was pending.
+  ///
+  /// If `isAlive` was disposed while the request was pending, will throw a [CancelledException].
+  Future<InstanceRef> safeEval(
+    String expression, {
+    @required Disposable isAlive,
+    Map<String, String> scope,
+  }) async {
+    Object result;
+
+    try {
+      if (disposed) {
+        throw StateError(
+          'Called `safeEval` on a disposed `EvalOnDartLibrary` instance',
+        );
+      }
+
+      result = await addRequest(isAlive, () async {
+        final libraryRef = await _waitForLibraryRef();
+
+        return await service.evaluate(
+          isolateId,
+          libraryRef.id,
+          expression,
+          scope: scope,
+        );
+      });
+
+      if (result == null) return null;
+
+      if (result is! InstanceRef) {
+        if (result is ErrorRef) {
+          throw EvalErrorException(
+            expression: expression,
+            scope: scope,
+            errorRef: result,
+          );
+        }
+        if (result is Sentinel) {
+          throw EvalSentinelException(
+            expression: expression,
+            scope: scope,
+            sentinel: result,
+          );
+        }
+        throw UnknownEvalException(
+          expression: expression,
+          scope: scope,
+          exception: result,
+        );
+      }
+    } catch (err, stack) {
+      _handleError(err, stack);
+      rethrow;
+    }
+
+    // We throw if the request was canceled instead of returning `null` because
```

Comment:
are you sure this is the case? You should get an InstanceRef with a null value not null. I think you might have been seeing null getting returned when eval was cancelled.

Reply:
Upon testing it, it appears that `eval('null')` does return an InstanceRef.

But `eval('[null]')` followed by a `getInstance` and reading `instance.elements.first` returns `null` (and I believe the issue is the same for maps)

So it appears to be inconsistent

Reply:
I've removed the null checks and updated the comment

Reply:
Sounds good. Lets file a bug to get the VM Service consistent on this point. I think you have discovered a bug in the VM Service api or a micro-optimization that makes the api less consistent.